cfpb/cfgov-refresh
# VideoPlayer consolidation and optional thumbnails

This PR encompasses a set of changes around our VideoPlayer module and related modules that include it:

1. VideoPlayers now are configured with YouTube video IDs (like [dQw4w9WgXcQ](https://www.youtube.com/watch?v=dQw4w9WgXcQ)) instead of YouTube embed URLs (like https://www.youtube.com/embed/dQw4w9WgXcQ).
2. VideoPlayers now support an optional thumbnail image which gets displayed before/after the video plays, instead of using the default YouTube video thumbnail. See internal el-camino#247.
3. The FeaturedContent organism now uses the VideoPlayer directly instead of using its own set of custom fields for video details (most of these fields were unused, see below).

EventPage usage of the VideoPlayer template has been updated to be consistent with these changes; however the support for alternate thumbnail images has not been added there.

This PR includes multiple Wagtail data migrations to properly alter existing video data. These migrations required some improvements to our [Wagtail StreamField migration](https://cfpb.github.io/cfgov-refresh/migrations/) code.

As part of this work, and to support easier debugging of video playback, I modified the JS code to support multiple VideoPlayers and FeaturedContent modules per page; previously the JS assumed a single module of each type.

A final minor improvement from this PR is that video playback will now work properly on Wagtail preview/draft views; previously our JS didn't work in those cases.

## How to test this PR

Running a local server with a production dump, here are some example pages to test:

1. VideoPlayer: http://localhost:8000/coronavirus/mortgage-and-housing-assistance/director-message-housing-resources/. Compare with behavior [in prod](https://www.consumerfinance.gov/coronavirus/mortgage-and-housing-assistance/director-message-housing-resources/).
2. FeaturedContent with video: http://localhost:8000/about-us/diversity-and-inclusion/self-assessment-financial-institutions/. Compare with behavior [in prod](https://www.consumerfinance.gov/about-us/diversity-and-inclusion/self-assessment-financial-institutions/).
3. EventPage with video: http://localhost:8000/about-us/events/archive-past-events/cfpb-financial-well-being-research-conference/. Compare with behavior [in prod](https://www.consumerfinance.gov/about-us/events/archive-past-events/cfpb-financial-well-being-research-conference/).

Experiment with setting custom thumbnails for these videos to see how they appear. For the EventPage changes, you can experiment with setting a live video ID for a live event to verify that it works properly.

(You can also use [the Wagtail Block Inventory tool](http://localhost:8000/admin/inventory/) to find other example pages with these modules.)

I've also added two new template debug views for VideoPlayer and FeaturedContent:

- http://localhost:8000/admin/template_debug/v1/video_player/
- http://localhost:8000/admin/template_debug/v1/featured_content/

## Screenshots

Configuring a VideoPlayer:

![image](https://user-images.githubusercontent.com/654645/84413020-11ec2700-abde-11ea-8859-c23803b3348e.png)

Video section in FeaturedContent:

![image](https://user-images.githubusercontent.com/654645/84412965-fda82a00-abdd-11ea-9293-a47dd8c1bb07.png)

New fields on EventPage:

![image](https://user-images.githubusercontent.com/654645/84413198-44961f80-abde-11ea-9177-a66bfaa53a9a.png)
![image](https://user-images.githubusercontent.com/654645/84413212-4a8c0080-abde-11ea-8863-d2c8fddaf791.png)

## Notes and todos

- I did some analysis of our FeaturedContent modules to make sure that we could safely transition this to using VideoPlayer. That analysis can be found [here](https://nbviewer.jupyter.org/gist/chosak/09f98263908f9d1335c6ce6605f12512) for anyone interested.
- This work exposed a few TODOs around FeaturedContent: for example, the body should be plain text, not rich text. These changes will be tackled in future PRs.
- If/when this is merged, I will update our internal Hubcap documentation around embedding videos.

## Checklist

_[Feel free to delete any checkboxes that are not applicable to this PR.]_

  - PR titles are used to generate the change log in [releases](../../releases),
    so good ones make that easier to scan.
  - Consider prefixing with the feature/section you're addressing,
    e.g., "Mega Menu: fix layout bug" or "Paying for College: Update content on Repay tool"

### Front-end testing

#### Browser testing

Visually tested in the following supported browsers:

[Further guidance on browser support](https://github.com/cfpb/development/blob/master/guides/browser-support.md)

#### Accessibility

- [x] Does not introduce new errors or warnings in
      [WAVE](https://wave.webaim.org/extension/)

#### Other
